### PR TITLE
Adding AC-GAN example on MNIST

### PR DESCRIPTION
**Pull request by lukedeo**
Provides a simple implementation of Auxiliary Classifier Generative Adversarial Networks from [this paper](https://arxiv.org/abs/1610.09585).

### CODE REVIEW

File: examples/mnist_acgan.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+"""
+from __future__ import print_function
+
+from collections import defaultdict
+import cPickle as pickle
+from PIL import Image
+
+from six.moves import range
+
+from keras.backend import set_image_dim_ordering
```
Comment (fchollet) [inline examples/mnist_acgan.py]: Prefer `from keras import backend as K` then call `K.*`.

File: examples/mnist_acgan.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+from keras.utils.generic_utils import Progbar
+import numpy as np
+
+np.random.seed(1337)
+
+set_image_dim_ordering('th')
+
+# nice, tabular formatting for progress messages
+HEADING = '{0:<22s} | {1:4s} | {2:15s} | {3:5s}'
+ROW_FMT = '{0:<22s} | {1:<4.2f} | {2:<15.2f} | {3:<5.2f}'
```
Comment (fchollet) [inline examples/mnist_acgan.py]: HEADING is only used in one place so does not need to be turned into a variable. ROW_FMT should be defined right before being used, rather than at the top of the file.

File: examples/mnist_acgan.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+
+def adam_config():
+    # parameters suggested in https://arxiv.org/abs/1511.06434, slows down
+    # Adam for better convergence
+    return Adam(lr=0.0002, beta_1=0.5, beta_2=0.999, epsilon=1e-08)
```
Comment (fchollet) [inline examples/mnist_acgan.py]: I would rather define the parameters of the optimizer as variables in the code, then instantiate `Adam` without this wrapper function.

File: examples/mnist_acgan.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    cnn.add(Dropout(0.3))
+
+    cnn.add(Flatten())
+
+    image = Input(shape=(1, 28, 28))
+
+    features = cnn(image)
+
+    # fake output tracks binary {fake, not-fake} discrimination, and the
+    # auxiliary requires reconstruction of latent features, in this case, labels
```
Comment (fchollet) [inline examples/mnist_acgan.py]: Please improve this comment, it's not immediately clear what it means.

**Comment by lukedeo** [comment]
squashed changes in, let me know if there's anything else.

**Comment by fchollet** [comment]
Great, thanks!

**Comment by bobchennan** [comment]
I have a question about the implementation.

According to the FAQ, after changing the ```trainable``` property we need to re-compile the model to take effect. But during the training process the model was not re-compiled so I am not sure during the generator training, the weights for descriminator changed or not.

Also we can try to use gradient reversal layer and compare the result. It did similar things but I am not sure which one is better

**Comment by farizrahman4u** [comment]
Weights of discriminator are not changed while training the combined model, since discriminator.trainable==False when the combined model was compiled.

**Comment by bobchennan** [comment]
Yes you are right.

So in fact Line 222 and Line 231 is just a reminder.

**Comment by farizrahman4u** [comment]
@bobchennan #4503

**Comment by appierys** [comment]
I have another question about the implementation. The `generator` model is not used directly for training. In this case, do we really need to compile the `generator` as done in this code? Does it have any differences if we omit the following line?

`generator.compile(optimizer=Adam(lr=adam_lr, beta_1=adam_beta_1), loss='binary_crossentropy')`

**Comment by bobchennan** [comment]
We need to use generator to generate fake samples. If I remember right if only prediction is needed we don't need to compile the model.
